### PR TITLE
Move json writing into separate classes

### PR DESCRIPTION
**Pull request by reinsch82**
This is to allow different implementations for problems to expose their specifics to the json that is put into the template.

We also moved configuration cache specifics out the the `CommonReport` class.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

Fixes https://github.com/gradle/gradle-private/issues/4267

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by hegyibalint — APPROVED** [review]
Apart from the Kotlin parts, LGTM.
I would wait for a second approval from the CC team regarding the Kotlin pieces.

File: platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy
Comment (reinsch82) [reply platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy:42]: I was using this test to try things and I didn't realize that it won't be part of the problem report, because the wrong api usage causes a distinct problem.
I'm hoping to make sure I remind that in the future.

File: platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy
Comment (reinsch82) [reply platforms/ide/problems-api/src/integTest/groovy/org/gradle/api/problems/ProblemsServiceIntegrationTest.groovy:428]: It's a copy of the above test to create distinct problems.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblemsSummary.kt
Comment (reinsch82) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblemsSummary.kt:24]: I moved `ProblemSeverity`, now it needs an explicit import

File: gradle.properties
Comment (reinsch82) [reply gradle.properties]: I can revert it, locally I have to have it to not run out of memory when importing in Intellij

Comment (ov7a) [reply gradle.properties]: I think that this is something that should be investigated at least. If you was hit by some changes, users will too

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-configuration/configuration-problems-base/build.gradle.kts
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/build.gradle.kts:47]: ❓  This is a big dependency, why do we need it now?

File: platforms/core-configuration/configuration-problems-base/build.gradle.kts
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/build.gradle.kts:60]: ❓  This is a big dependency, why do we need it now?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/HtmlReportWriter.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/HtmlReportWriter.kt]: 💭  Let's introduce a data class with helpful field names

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:17]: 🤔  Did we lose a package declaration?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt:19]: ❌  This class has wrong package

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt:56]: 💅  Let's use an existing utility instead

https://github.com/gradle/gradle/blob/c80baa2d7bf3fcb0740176bebeabe35c2e3852d7/platforms/core-configuration/stdlib-kotlin-extensions/src/main/kotlin/org/gradle/internal/extensions/stdlib/CharSequenceExtensions.kt#L20

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/DecoratedReportProblem.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/DecoratedReportProblem.kt:36]: 💅  Kotlin classes are final by default

Also, does the `problem` need to be public?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/DocumentRegistryExtensions.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/DocumentRegistryExtensions.kt:2]: ❔

File: platforms/ide/problems/build.gradle.kts
Comment (alllex) [inline platforms/ide/problems/build.gradle.kts:63]: 💅  Swap the two lines for perfect ordering

Thanks for ordering the rest!

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:52]: 🤔  Why not change the types in the constructor to be non-nullable?

In the other place, we use a space-separated string for `"configuration cache report"`. We should probably make sure this is consistent.

File: testing/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java
Comment (alllex) [inline testing/architecture-test/src/test/java/org/gradle/architecture/test/ArchUnitFixture.java:90]: 🤔  Are all classes in the specified package written exclusively in Kotlin? If not, we risk losing coverage in all tests that use this predicate

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:85]: 🤔  Is this safe under concurrency?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt:256]: 🤔  Maybe we shouldn't mix the concerns? I think the report file name was always intended to have dashes, and now it's not

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:66]: 🤔  If this is some sort of identifier, I'd avoid using spaces there.

Also, we seem to be inconsistent with "problemS report" vs "problem report" in other places

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:90]: 💭  I am starting to doubt the current approach is maintainable. It's becoming rather fragile.

Previously, there was a single place where we've written the data in the JSON format, which was effectively the contract expected by the JavaScript part. Now, this is at least in two different places.

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:66]: ❌  Having this marker implies that the extra data expected by the problems report will be at the top level of the JSON as well. I think it makes it harder to maintain and review.

Instead of mixing the data, let's introduce a single top-level field in the JSON that will carry all the extra data needed for the Problems Report. The presence of that field will indicate to the JS report logic that it's a Problems Report rather than CC report, and it can proceed with processing this data.

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:73]: ❌  There should be no need in writing CC-specific data to the model.

If this is done to accommodate the requirements on the data conversion on the report side, I suggest we merge the changes required on the report side first, before proceeding with the PR on the `gradle/gradle` side.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/ProblemReportDetails.kt
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/ProblemReportDetails.kt:38]: ❓  Why do we have the top-level data writing logic duplicated here and in `DefaultProblemsReportCreator` ?

File: platforms/core-configuration/configuration-problems-base/build.gradle.kts
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/build.gradle.kts:60]: for BuildNameProvider -> SettingsInternal

File: platforms/core-configuration/configuration-problems-base/build.gradle.kts
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/build.gradle.kts:47]: I think `core`drags that in.

```
Execution failed for task ':configuration-problems-base:projectHealth'.
> These transitive dependencies should be declared directly:
    api(project(":core-api"))

```

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:17]: indeed we did :|

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/CommonReport.kt:256]: Do you mean we should have a separate parameter for filename and context string?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/ProblemReportDetails.kt
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/ProblemReportDetails.kt:38]: It won't stay duplicated once we finalize what actual data we need to transfer.

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (reinsch82) [reply platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:66]: I'm all for it, but can we do that later when we ironed out what we exactly want to transfer into the json?
The purpose for now is the separation, so we can do different things for CC and problems.

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (reinsch82) [reply platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:90]: As of now there will be 2 contracts.
Therefore it can obviously not stay in the same place it was.
Maybe a collection of extension methods per "contract" could avoid the separation?

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (reinsch82) [reply platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt]: This will be different soon.

I intend to use an objects presence to indicate the type of report not the a string.

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/DefaultProblemsReportCreator.kt:55]: 💅 This does not have to be a `var`

File: platforms/ide/problems/build.gradle.kts
Comment (alllex) [inline platforms/ide/problems/build.gradle.kts:19]: I double checked and all modules that apply this `implementation-kotlin` plugin have sources in `main` exclusively in Kotlin. Given that this is just for one class and the potential complications with the ArchUnit test, I would suggest rewriting this one class in Java.

File: .teamcity/src/main/kotlin/common/extensions.kt
Comment (reinsch82) [reply .teamcity/src/main/kotlin/common/extensions.kt:307]: I replaced the usages with

`org.gradle.internal.extensions.stdlib.CharSequenceExtensionsKt#capitalized`

File: platforms/ide/problems/build.gradle.kts
Comment (reinsch82) [reply platforms/ide/problems/build.gradle.kts:19]: I'll move the class to a different package.

Reimplementing it has also more complications than just this class. e.g. extension methods.

**Comment by reinsch82** [comment]
@bot-gradle test this

File: .teamcity/src/main/kotlin/common/extensions.kt
Comment (alllex) [reply .teamcity/src/main/kotlin/common/extensions.kt:307]: That would be even more surprising, given that `.teamcity` is a Maven project that does not depend on Gradel production sources, AFAIK

File: platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt
Comment (alllex) [inline platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt:61]: ❌  This does not seem right

Comment (reinsch82) [reply platforms/ide/problems/src/main/kotlin/org/gradle/problems/internal/impl/DefaultProblemsReportCreator.kt:61]: how did that happen?!

File: platforms/ide/problems/build.gradle.kts
Comment (alllex) [reply platforms/ide/problems/build.gradle.kts:19]: I guess that could work, and we can merge it like this. Though, I am still a bit skeptical about mixing Kotlin with Java. It should work, but I don't know if there was a reason why we never did it.

The ArchUnit is also something we need to improve, because it becomes too easy to include packages that contain both Kotlin and Java.

**Review by alllex — APPROVED** [review]

**Comment by reinsch82** [comment]
@bot-gradle test this

File: .teamcity/src/main/kotlin/common/extensions.kt
Comment (cobexer) [reply .teamcity/src/main/kotlin/common/extensions.kt:307]: TeamCity DSL does not compile after this: https://builds.gradle.org/buildConfiguration/Hygiene_TestTeamCityBuildConfigurations/85632262?buildTab=log&linesState=82.555&logView=flowAware&focusLine=543

* https://github.com/gradle/gradle/pull/30009

**Pull request merged**
